### ISSUE TITLE
Feature request - incoming transaction notifications

### ISSUE THREAD

**Issue by szollo** [cryptoadvance/specter-desktop#778]
It would be amazing to have support for desktop notifications of incoming transactions, so that the user is informed when there's new incoming bitcoin without needing to refresh the browser, or indeed even look at the web interface - this is pretty much the only feature preventing me to fully switch from Electrum at this point.

Taking a cursory look this could be achieved by using ZMQ subscriptions, however as it's not plug-and-play it would probably have to be an advanced setting, as Bitcoin Core default config doesn't enable it (though it's compiled into the official binaries afaik).

The other challenges I can think of, off the top of my head:

- Having cross-platform desktop notification (perhaps just browser notifications instead of OS native would suffice?)
- Extra configuration steps if Specter is connecting to the node via Tor, seeing as the ZMQ port would also need to be published as a separate hidden service
- There's probably a client / frontend websocket component to develop?

Thanks for the great work, Specter is awesome.

**Comment by relativisticelectron** [comment]
I thought the enhancement is a good idea and developed a proof-of-concept for it: https://github.com/relativisticelectron/specter-desktop/tree/notification

**Comment by relativisticelectron** [comment]
@moneymanolis : Do you think this web-push notification would be useful for many users? Or are most users running bitcoin-qt locally, and have therefore the notifications from bitcoin-qt?

**Comment by szollo** [comment]
That's exactly what I had in mind @relativisticelectron - looks great!

**Comment by moneymanolis** [comment]
@relativisticelectron looks cool! As for your question about bitcoin-qt, I think nobody uses this, users have usually different variations of bitcoind instances running, either locally or on a remote node. 

Tagging @k9ert for architectural input.  

Sth. that directly caught my eye from the MDN documentation:

<img width="808" alt="image" src="https://user-images.githubusercontent.com/70536101/156052967-015f2746-c58d-422a-922f-ddfdc5d45043.png">
Don't we have a user group that struggles with not being able to connect via HTTPS with their (remote) node? I think Umbrel, no?

**Comment by relativisticelectron** [comment]
> <img alt="image" width="808" src="https://user-images.githubusercontent.com/70536101/156052967-015f2746-c58d-422a-922f-ddfdc5d45043.png">
> Don't we have a user group that struggles with not being able to connect via HTTPS with their (remote) node? I think Umbrel, no?

Yes, I have no idea how to deal with this.   It seems https is not coming to umbrel any time soon: https://github.com/getumbrel/umbrel-os/issues/251     ...

**Comment by k9ert** [comment]
Wow, this really looks cool. Somehow overlooked this issue. I think the zeromq thing should be more in focus than the frontend-implementation as otherwise, you just create unnecessary core-calls which might cause unnecessary loads on small raspis.

**Comment by relativisticelectron** [comment]
Added zeroqm such that rpc calls for notifications are ONLY done if the zeroqm has gotten a message:  https://github.com/cryptoadvance/specter-desktop/compare/master...relativisticelectron:notification?expand=1

**Comment by k9ert** [comment]
Wow, this really looks promising and also easier than i thought. When i get that right, you don't clear the buffer to the server-side data-structres directly but just create the notification on the client-side (via continuous pulling and maybe later via web-push) and the then the user is encouraged to make a normal request requests the data. We probably need to also put some logic in for the case: It should profit also via not asking core that often anymore but all of that can wait for now.

Not sure how complex the web-push is but maybe we can separate the two things? I'd probably hesitate to merge web-push if it's not working on firefox but who cares for cheap and open small polls?!

I just had a look in the internal node implementation. Indeed, we have a `bitcoin.conf`. If we modify that one, we would need a `SpecterMigration`. But i don't think it's necessary. I don't think we should have created a `bitcoin.conf` in the first place and another alternative is to "add" the zmq-config via the commandline here:
https://github.com/cryptoadvance/specter-desktop/blob/master/src/cryptoadvance/specter/process_controller/node_controller.py#L333

Can you try to add the necessary ZMQ-stuff into the commandline? Maybe create an extra `config.py` config value which makes it possible to deactivate ZMQ.  IMHO, the commandline options come on top of the `bitcoin.conf`.

**Comment by relativisticelectron** [comment]
Thanks for the feedback!  

- Great that you think that the simple javascript `setInterval` checker is sufficient. The checker checks [if any messages are in the zmq buffer](https://github.com/relativisticelectron/specter-desktop/blob/ce2d5857913d4ab370603359857e46d6709e513d/src/cryptoadvance/specter/server_endpoints/wallets_api.py#L106).  
- Sidenote:  I first thought pywebpush would work also in firefox, but after I got it working was disappointed that it only works in Chrome.  Additionally I don't like the idea of sending the (encrypted) notifications through the google cloud --  so I move the pywebpush code onto a [different (trash) branch](https://github.com/relativisticelectron/specter-desktop/tree/notifications_pywebpush).
- Yes I will put the zmq it in the commandline options and test it soon.

**Comment by relativisticelectron** [comment]
If the permission cannot be granted (because of https missing), and is not denied I can at least show javascript notifications:
![Peek 2022-04-20 21-52](https://user-images.githubusercontent.com/60378539/164311894-bb8b9e12-d002-4bba-9282-562103534e8b.gif)

Would these be helpful?

**Comment by szollo** [comment]
That looks great @relativisticelectron!

> @szollo : Would these Javascript notifications (in the browser window) be helpful if the [Notifications_API](https://developer.mozilla.org/en-US/docs/Web/API/Notifications_API/Using_the_Notifications_API) ([docs](https://notifications.spec.whatwg.org/#dom-notification)) is not available?

For the use case I had in mind I'd prefer it was native to the OS (ie. macOS or Windows notifications) or browser, as more often than not I don't view the active window, especially when travelling and using one screen.

That being said, it's great to have the choice - so if I wanted to have that feature I'd have to make sure I run Specter as an HTTPS service with a trusted certificate?

**Comment by relativisticelectron** [comment]
> For the use case I had in mind I'd prefer it was native to the OS (ie. macOS or Windows notifications) or browser, as more often than not I don't view the active window, especially when travelling and using one screen.

Sure, that is best. Without https this Notification API (remotely)  is just not possible. I still have to investigate if the push service (google or [mozilla](https://blog.mozilla.org/services/2016/08/23/sending-vapid-identified-webpush-notifications-via-mozillas-push-service/)) works without https, but it would involve a cloud server that forwards encrypted messages.
TODO:

- [ ] test pywebpush via remote Specter
- [ ] if it works (yeah!!!), try the [mozilla](https://blog.mozilla.org/services/2016/08/23/sending-vapid-identified-webpush-notifications-via-mozillas-push-service/) cloud server 

> That being said, it's great to have the choice - so if I wanted to have that feature I'd have to make sure I run Specter as an HTTPS service with a trusted certificate?

1. [Notifications_API](https://developer.mozilla.org/en-US/docs/Web/API/Notifications_API/Using_the_Notifications_API) ([docs](https://notifications.spec.whatwg.org/#dom-notification))  on Localhost doesn't require a certificate.
2. Yes for remote specter, [Notifications_API](https://developer.mozilla.org/en-US/docs/Web/API/Notifications_API/Using_the_Notifications_API) ([docs](https://notifications.spec.whatwg.org/#dom-notification)) does require a certificate. But since this is probably not available in most node packages, maybe javascript notifications are better than no notifications at all?

**Comment by k9ert** [comment]
Again missed your comment even though in the meantime i'm indeed using the github-task/notification system.

> I that OK for everyone @k9ert , @moneymanolis @szollo ? Should I (slowly) develop this [proof-of-concept]
> (https://github.com/relativisticelectron/specter-desktop/tree/notification) into a PR?

I think this looks very promising. Maybe we can somehow combine that with #1695 ? On the browser-side, I'd focus for now on the Javascript Notifications and defer the notification-API to the very end. So imho the order should be: 
* basic notification system (maybe with some pages to read/archive)
* ZMQ connection integrating with the notification-system
* Javascript Notifications
* Notifications API

Thank you so much, you're doing a great work!
